libs/java/gcp_zts_creds/src/main/java/com/yahoo/athenz/creds/gcp/GCPSIACredentials.java

```diff
@@ -0,0 +1,270 @@
+import org.apache.http.HttpEntity;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.client.methods.RequestBuilder;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.bouncycastle.asn1.DERIA5String;
+import org.bouncycastle.asn1.x509.GeneralName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.security.PrivateKey;
+import java.security.cert.X509Certificate;
+
+public class GCPSIACredentials {
+
+    // Configurations.
+    public static int ZTS_CONNECT_TIMEOUT_MS = 5000;
+    public static int ZTS_READ_TIMEOUT_MS = 30000;
+    public static int ATTESTATION_CONNECT_TIMEOUT_MS = ZTS_CONNECT_TIMEOUT_MS;
+    public static int ATTESTATION_READ_TIMEOUT_MS = ZTS_READ_TIMEOUT_MS;
+
+    /** Response of {@link #getGCPFunctionServiceCertificate} */
+    public static class PrivateAndCertificate {
+        public X509Certificate certificate;
+        public String          certificatePem;
+        public PrivateKey      privateKey;
+        public String          privateKeyPem;
+        public String          caCertificatesPem;
+    }
+
+    /**
+     * For GCP cloud-functions generate a new private key, request an
+     * x.509 certificate based on the requested CSR and return both to
+     * the client in order to establish tls connections with other
+     * Athenz enabled services.
+     * @param athenzDomain name of the domain
+     * @param athenzService name of the service
+     * @param gcpProjectId GCP project-id that the function runs in
+     * @param athenzProvider name of the provider service for GCP Cloud-Functions
+     * @param ztsUrl Something like: https://...:.../zts/v1
+     * @param certDomain TODO: Abhijeet - explain what this is...
+     * @param optionalCountry Optional field in the certificate's Subject.
+     * @param optionalState Optional field in the certificate's Subject.
+     * @param optionalLocality Optional field in the certificate's Subject.
+     * @param optionalOrganization Optional field in the certificate's Subject.
+     * @param optionalOrganizationUnit Optional field in the certificate's Subject.
+     * @return GCPFunctionIdentity with private key and certificate
+     */
+    public static PrivateAndCertificate getGCPFunctionServiceCertificate(
+            String athenzDomain,
+            String athenzService,
+            String gcpProjectId,
+            String athenzProvider,
+            String ztsUrl,
+            String certDomain,
+            String optionalCountry,
```

Comment:
we should not include optional in the field names to identity the field as optional - these are already included in the javadocs. let's just include rdn instead of optional to identity that these are relative distinguished name components - e.g. rdnCountry, rdnState, etc.

Reply:
👍